### ISSUE TITLE
nitcorn paths

### ISSUE THREAD

**Issue by Morriar** [nitlang/nit#2030]
Hello everyone, 

Until now, nitcorn is very permissive regarding url matching.

If we take the route `/user`, here the matching of diverse routes:

~~~
false /
true  /user
true  /user/Morriar
true  /user/Morriar/profile
~~~

The same apply for vararg routes, for example with `/user/:id`

~~~
false /
false  /user
true  /user/Morriar
true  /user/Morriar/profile
~~~

Since the route matching of nitcorn is based on the order of declaration of the routes, the following route/action declaration will always be redirected to `IndexAction`.

~~~
/ => IndexAction
/user => UserListAction
/user/Morriar => UserAction
/user/Morriar/profile => UserProfileAction
~~~

This can drives the developer crazy when handling 10+ routes...

I suggest to tighten the route matching a bit here the proposed spec:

* `/` maches only `/`
* `/user` maches only `/user`
* `/user/:id` maches only `/user/{someid}` but not `/user/{someid}/profile`
* `/user/:id/profile` maches only `/user/{someid}/profile`

To make the migration easier and also because it is really useful, I propose a new kind of route suffixed by `*`.

* `/*` maches everything (the current behavior of `/`)
* `/user/*` maches `/user/{id}/` and `/user/{id}/profile` (not sure about `/user`, if I look at the current behavior of nitcorn regarding `/user` and `/user/`, I would go for `/user` is not matched)
* `/user/:id/profile/*` maches `/user/{someid}/profile/foo/bar`

I can submit a pull request that makes the change in the lib. @xymus @R4PaSs what do you think regarding you clients?

**Comment by Morriar** [comment]
Also, what about the matching priority?

* for the request `user/profile` how to choose between routes `/user/:id` and `/user/:id`
* for the request `user/profile` how to choose between routes `/user/` and `/user/*`

**Comment by xymus** [comment]
Instead of adding `*` to existing routes, you should use the end marker `$` for the new non-trailing routes. Changing the current behavior would break many clients. Opportunity, Tnitter, Benitlux and most of the file servers rely entierly on routes with a trailing part.

It would serve the same purpose, the end marker could point index without catching everything using `/$ -> IndexAction`, and you could lock those varargs to a specific length.

For the matching priority, I would keep the declaration order to break ties. Leaving the user choose allows for more customization than forcing one or the other in the lib.

**Comment by Morriar** [comment]
Yeah I tough of the `$` solution then I asked myself what should be the default?

In fact the choice here is: permissive matching vs strict matching by default. 

I made some tests:

* http://google.ca/foo => 404
* https://github.com/foo/bar/baw => 404 (there is actually a `foo` user so `/foo`) works :p)
* https://www.reddit.com/foo => 404

Even the traditional outsiders seems to comply:
* http://uqam.ca/foo => 404
* https://www.microsoft.com/foo => 404

And then:
* http://opportunity.gresil.org/foo/bar/bax/i/m/lost => 200 :dancer: 

Then I looked at some web framework to see the default behavior, and most of them seem to follow the strict route matching rule:

* express.js
* ember.js
* sinatrarb
* rails
* appengine
* angular (does that with the `then` route)

I still more in favor of the strict matching route by default and let the user activate the permissive mode by hand with `*`.

**Comment by xymus** [comment]
We should not skip steps, first we add the strict routes as an option. And in the future, if it becomes widely used, we can see about making it the default

Note that in Opportunity we use the index as a catch all, that's the desired behavior. However all nitcorn REST APIs (that I know of) return a custom 400 or 404 on invalid requests:
* http://xymus.net/foo gives a 404
* Tnitter API: http://tnitter.xymus.net/rest/ 400
* Benitlux API at, old: http://xymus.net/benitlux/rest/ and new: http://xymus.net:8080/rest/foo 400
* Even the API interface of Opportunity is clean, though silent to a human: http://xymus.net/opportunity/rest/foo 400

The strict matching would be useful, but it's not worth changing the default behavior. I would use strict matching on index, but that's pretty much it, for APIs and Actions handling more than one resource, permissive matching is more useful. I think nitcorn differs from other libs because each route is associated to a class instead of a method, so we tend to use less routes and larger actions. This may change in the future, but that's what we are working with today.

How about making Route use a Pattern instead of a String?

**Comment by privat** [comment]
closed by #2094
